core/miner/storage_fsm/impl/precommit_batcher_impl.cpp

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "miner/storage_fsm/impl/precommit_batcher_impl.hpp"
+
+#include <utility>
+#include "const.hpp"
+#include "vm/actor/builtin/v0/miner/miner_actor.hpp"
+namespace fc::mining {
+  using api::kPushNoSpec;
+  using fc::vm::actor::builtin::v0::miner::PreCommitBatch;
+  using libp2p::protocol::scheduler::toTicks;
+  using primitives::ChainEpoch;
+  using vm::actor::builtin::types::miner::kChainFinality;
+  PreCommitBatcherImpl::PreCommitBatcherImpl(const Ticks &max_time,
+                                             std::shared_ptr<FullNodeApi> api,
+                                             const Address &miner_address,
+                                             const Ticks &closest_cutoff)
+      : max_delay_(max_time),
+        api_(std::move(api)),
+        miner_address_(miner_address),
+        closest_cutoff_(closest_cutoff){};
+
+  PreCommitBatcherImpl::PreCommitEntry::PreCommitEntry(
+      TokenAmount number, api::SectorPreCommitInfo info)
+      : deposit(std::move(number)), precommit_info(std::move(info)){};
+
+  outcome::result<std::shared_ptr<PreCommitBatcherImpl>>
+  PreCommitBatcherImpl::makeBatcher(
+      const Ticks &max_wait,
+      std::shared_ptr<FullNodeApi> api,
+      std::shared_ptr<libp2p::protocol::Scheduler> scheduler,
+      const Address &miner_address) {
+    std::shared_ptr<PreCommitBatcherImpl> batcher =
+        std::make_shared<PreCommitBatcherImpl>(
+            max_wait, std::move(api), miner_address, max_wait);
+
+    batcher->cutoff_start_ = std::chrono::system_clock::now();
+    batcher->logger_ = common::createLogger("batcher");
+    batcher->logger_->info("Bather have been started");
+    batcher->handle_ = scheduler->schedule(max_wait, [=]() {
+      batcher->sendBatch();
+      batcher->handle_.reschedule(max_wait);  // TODO: maybe in gsl::finally
+    });
+    return batcher;
+  }
+
+  void PreCommitBatcherImpl::sendBatch() {
+    std::unique_lock<std::mutex> locker(mutex_, std::defer_lock);
+    // TODO(Elestrias): [FIL-398] goodFunds = mutualDeposit + MaxFee; -  for
+    // checking payable
+
+    if (batch_storage_.size() != 0) {
+      logger_->info("Sending procedure started");
+      auto head = api_->ChainHead().value();
+      auto minfo = api_->StateMinerInfo(miner_address_, head->key).value();
+
+      PreCommitBatch::Params params = {};
+      for (const auto &data : batch_storage_) {
+        mutual_deposit_ += data.second.deposit;
+        params.sectors.push_back(data.second.precommit_info);
+      }
+      auto encodedParams = codec::cbor::encode(params);
+      if (encodedParams.has_error()) {
+        logger_->error("Error has occurred during parameters encoding: {}",
+                       encodedParams.error().message());
```

Comment:
I think that it is necessary to break the method if we meet an error.
On this moment we meet an error, log it and continue the function.
It will proceed segfault above when we try to call encodedParams.value() below.
It is related to all `var.has_error()` checks
@ortyomka what is expected behaviour in such cases?

Reply:
we decided to make a vector of callbacks and change the signature of void SendBatch() to outcome::result<void> sendBatch(PrecommitBatchResult & precommit_batch_result); to inform sealing processes about the results of the sending